### PR TITLE
Adding fdatasync and fsync system calls

### PR DESCRIPTION
**Pull request by kuzeyardabulut**

### CODE REVIEW

**Review by RusherRG — APPROVED** [review]
Nice fix!!
This should help us solve the data persistence issue with Postgres.

File: src/safeposix/syscalls/fs_calls.rs [hunk 1]
```diff
@@ -1843,6 +1843,72 @@ impl Cage {
...
+    //------------------------------------FSYNC SYSCALL------------------------------------
+
+    pub fn fsync_syscall(&self, fd: i32) -> i32 {
+        let checkedfd = self.get_filedescriptor(fd).unwrap();
+        let mut unlocked_fd = checkedfd.write();
+        if let Some(filedesc_enum) = &mut *unlocked_fd {
+            match filedesc_enum {
+                File(ref mut normalfile_filedesc_obj) => {
+                    if is_rdonly(normalfile_filedesc_obj.flags) {
+                        return syscall_error(Errno::EBADF, "fsync", "specified file not open for sync");
```
Comment (RusherRG) [inline src/safeposix/syscalls/fs_calls.rs:1855]: Can we not sync read only files or are we choosing not to sync them?

Comment (rennergade) [reply src/safeposix/syscalls/fs_calls.rs:1855]: nothing can be written to them so they shouldnt be synced

File: src/safeposix/syscalls/fs_calls.rs [hunk 2]
```diff
@@ -1843,6 +1843,72 @@ impl Cage {
...
+                            syscall_error(Errno::EROFS, "fsync", "does not support special files for synchronization")
+                        }
+                    }
+                }
+                _ => {syscall_error(Errno::EINVAL, "fsync", "fd is attached to an object which is unsuitable for synchronization")}
+            }
+        } else {
+            syscall_error(Errno::EBADF, "fsync", "invalid file descriptor")
+        }
+    }
```
Comment (yzhang71) [inline src/safeposix/syscalls/fs_calls.rs:1877]: The combination of if let and nested match statements can be a bit hard to follow. Consider refactoring to make the logic clearer. Sometimes, flattening nested matches or replacing them with other control flow structures can enhance readability.

File: src/safeposix/syscalls/fs_calls.rs [hunk 3]
```diff
@@ -1843,6 +1843,72 @@ impl Cage {
...
+
+    //------------------------------------FDATASYNC SYSCALL------------------------------------
+
+    pub fn fdatasync_syscall(&self, fd: i32) -> i32 {
+        let checkedfd = self.get_filedescriptor(fd).unwrap();
```
Comment (yzhang71) [inline src/safeposix/syscalls/fs_calls.rs:1882]: Avoiding unwrap(): The usage of unwrap() on get_filedescriptor and when retrieving inodeobj and fileobject can cause panics if the result is None. It's better to handle these cases with error propagation or more graceful error handling to avoid abrupt termination of the program.

File: src/safeposix/syscalls/fs_calls.rs [hunk 4]
```diff
@@ -1843,6 +1843,72 @@ impl Cage {
...
+                        }
+                    }
+                }
+                _ => {syscall_error(Errno::EINVAL, "fdatasync", "fd is attached to an object which is unsuitable for synchronization")}
+            }
+        } else {
+            syscall_error(Errno::EBADF, "fdatasync", "invalid file descriptor")
+        }
+    }
+
```
Comment (yzhang71) [inline src/safeposix/syscalls/fs_calls.rs:1911]: The repeated use of syscall_error can be refactored. Consider creating a helper function that takes parameters like Errno, error context, and message to construct and return the syscall error. This will reduce duplication and make the code cleaner.
```
enum FileSyncError {
    InvalidFileDescriptor,
    ReadOnly,
    IOError,
    // ... other error types
}
```
